File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/design/factory/simple/VehicleFactory.scala
package org.lang.scala.conclusion.design.factory.simple

import org.lang.scala.conclusion.design.factory.simple.common.vehicle.parent.Vehicle
import org.lang.scala.conclusion.design.factory.simple.common.vehicle.Ship
import org.lang.scala.conclusion.design.factory.simple.common.vehicle.Train
import org.lang.scala.conclusion.design.factory.simple.common.vehicle.Truck

/**
 * 	This is a class to implement simple factory pattern
 *  -- Declare a factory method to return an instance of different classes according to different arguments 
 *     -- Make use of polymorphism inside the factory method
 *     
 * 	@author VinceYuan
 */
object VehicleFactory {
 
  /**
   * 	This is a method to get vehicle from a type 
   */
  def getVehicle(`type`: String) = {
 
    /*	Declare an reference of parent interface	*/
    var vehicle: Vehicle = null
    
    /*	Reference to an instance of different child classes according to the argument	 */
    `type`.toUpperCase() match {
      case "SHIP" => vehicle = new Ship()
      case "TRAIN" => vehicle = new Train()
      case "TRUCK" => vehicle = new Truck()
    }
    
    /*	Return the instance  */
    vehicle
  }
}

File: JavaAndScala/src/main/java/com/java/se/conclusion/design/factory/simple/VehicleFactory.java
package com.java.se.conclusion.design.factory.simple;

import com.java.se.conclusion.design.factory.simple.common.vehicle.Ship;
import com.java.se.conclusion.design.factory.simple.common.vehicle.Train;
import com.java.se.conclusion.design.factory.simple.common.vehicle.Truck;
import com.java.se.conclusion.design.factory.simple.common.vehicle.Vehicle;

/**
 * 	This is a class to implement simple factory pattern
 *  -- Declare a public and static factory method to return an instance of different classes according to different arguments 
 *     -- Make use of polymorphism inside the factory method
 * @author VinceYuan
 *
 */
public class VehicleFactory {

	/**
	 * 	This is a public and static factory method
	 *  -- Declare an reference of parent abstract class
	 *  -- Reference to an instance of different child classes according to the argument
	 *  -- Return the instance
	 * @param type
	 * @return
	 */
	public static Vehicle getVehicle(String type) {
		
		/*	Declare an reference of parent abstract class	*/
		Vehicle vehicle;
		
		/*	Reference to an instance of different child classes according to the argument	*/
		switch (type.toUpperCase()) {
		case "SHIP":
			vehicle = new Ship();
			break;
		case "TRAIN":
			vehicle = new Train();
			break;
		case "TRUCK":
			vehicle = new Truck();
			break;
		default:
			vehicle = null;
			break;
		}
		
		/*	Return the instance	 */
		return vehicle;
	}
}
